### PR TITLE
setting up pubsub for room settings

### PR DESCRIPTION
**Pull request by BrandonMurnane**

### CODE REVIEW

File: config/redisConfig.js [hunk 1]
```diff
@@ -36,8 +36,14 @@ module.exports = {
     /*
      * PubSub for real-time events.
      */
-    pubsub: pe.REDIS_PUBSUB && pe[pe.REDIS_PUBSUB] ?
-      pe[pe.REDIS_PUBSUB] : PRIMARY_REDIS,
+    pubsubPerspective: pe.REDIS_PUBSUB_PERSPECTIVES && pe[pe.REDIS_PUBSUB_PERSPECTIVES] ?
+      pe[pe.REDIS_PUBSUB_PERSPECTIVES] : PRIMARY_REDIS,
+
+    /*
+     * PubSub for real-time events.
```
Comment (tausifmuzaffar) [inline config/redisConfig.js]: update comment for bots

File: realtime/redisSubscriber.js [hunk 2]
```diff
@@ -11,7 +11,7 @@
  */
 'use strict'; // eslint-disable-line strict
 const emitter = require('./socketIOEmitter');
-const sub = require('../cache/redisCache').client.sub;
+const subPerspective = require('../cache/redisCache').client.subPerspective;
```
Comment (tausifmuzaffar) [inline realtime/redisSubscriber.js]: is perspective a good work because refocus already uses that word to mean something else?

File: tests/realtime/realtimeUtils.js [hunk 3]
```diff
@@ -132,6 +141,12 @@ describe('tests/realtime/realtimeUtils.js, realtime utils Tests >', () => {
         expect(realtimeUtils.shouldIEmitThisObj(nspString,
           looksLikeSampleObjNAUS)).to.equal(true);
       });
+
+      it('for should return true for roomTest', () => {
```
Comment (tausifmuzaffar) [inline tests/realtime/realtimeUtils.js]: there should be fail test

File: realtime/redisSubscriber.js (hunk 2 above)
Comment (BrandonMurnane) [reply realtime/redisSubscriber.js]: that was ians suggestion he wanted to avoid it being refocus/imc

Comment (tausifmuzaffar) [reply realtime/redisSubscriber.js]: Okay cool, long as it got permission then sounds good

File: cache/redisCache.js [hunk 4]
```diff
@@ -58,17 +58,22 @@ if (featureToggles.isFeatureEnabled('enableRedisConnectionLogging')) {
   logger.info('Redis Retry Strategy', opts);
 }
 
-const sub = redis.createClient(rconf.instanceUrl.pubsub, opts);
-sub.subscribe(rconf.channelName);
+const subPerspective = redis.createClient(rconf.instanceUrl.pubsubPerspective, opts);
+subPerspective.subscribe(rconf.channelName);
+
+const subBot = redis.createClient(rconf.instanceUrl.pubsubBots, opts);
+subBot.subscribe(rconf.channelName);
```
Comment (iamigo) [inline cache/redisCache.js]: I know I said channel name didn't really matter but upon further reflection I think we should use a separate channel name for the bot pub/sub

File: db/helpers/common.js [hunk 5]
```diff
@@ -137,6 +139,17 @@ function createDBLog(inst, eventType, changedKeys, ignoreAttributes) {
 }
 
 /**
+ * A function to see if an object is a room object or not. It returns true
+ * if an object passed has 'type' and 'settings' as its property.
+ * @param  {Object}  obj - An object instance
+ * @returns {Boolean} - returns true if the object has the property
+ * "type" && "settings"
```
Comment (iamigo) [inline db/helpers/common.js]: This doesn't seem like a particularly strong type check. I *think* when you actually get a sequelize model instance, you should have a reference to the model itself. Can you check that?

File: db/helpers/common.js [hunk 6]
```diff
@@ -137,6 +139,17 @@ function createDBLog(inst, eventType, changedKeys, ignoreAttributes) {
...
+ */
+function isThisRoom(obj) {
```
Comment (iamigo) [inline db/helpers/common.js]: also, functions which return boolean should be named as positive assertion, so this should be "isRoom"

File: realtime/constants.js [hunk 7]
```diff
@@ -37,6 +39,12 @@ module.exports = {
       del: 'refocus.internal.realtime.subject.remove',
     },
 
+    room: {
+      add: 'imc.internal.realtime.room.add',
```
Comment (iamigo) [inline realtime/constants.js]: is the "imc" name surviving the jump into refocus? i think these should be refocus.internal.realtime.room...

File: realtime/socketIOEmitter.js [hunk 8]
```diff
@@ -26,8 +26,8 @@ module.exports = (io, key, obj) => {
 
   for (const nsp in io.nsps) {
     // Send events only if namespace connections > 0
-    if (nsp && Object.keys(nsp).length &&
-         rtUtils.shouldIEmitThisObj(nsp, obj)) {
+    if (nsp && (Object.keys(nsp).length > 0) &&
```
Comment (iamigo) [inline realtime/socketIOEmitter.js]: why?

File: realtime/utils.js [hunk 9]
```diff
@@ -39,6 +39,17 @@ function isThisSubject(obj) {
...
 
 /**
+ * A function to see if an object is a room object or not. It returns true
+ * if an object passed has 'type' and 'settings' as its property.
+ * @param  {Object}  obj - An object instance
+ * @returns {Boolean} - returns true if the object has the property
+ * "type" && "settings"
+ */
+function isThisRoom(obj) {
+  return obj.hasOwnProperty('type') && obj.hasOwnProperty('settings');
```
Comment (iamigo) [inline realtime/utils.js]: again, feels like weak type check
again, function name
here, i guess you're not necessarily guaranteeing the object to be an instance of a sequelize model? what could you do to tighten the type check?

File: realtime/utils.js [hunk 10]
```diff
@@ -211,7 +223,7 @@ function shouldIEmitThisObj(nspString, obj) {
      * subjectAbsolutePath in it, so we do not have to check for the filter
      * conditions and we just need to return true.
      */
-    if (nspComponents.length < 2) {
+    if (nspComponents.length <= 2) {
```
Comment (iamigo) [inline realtime/utils.js]: Why this change?

File: realtime/utils.js [hunk 11]
```diff
@@ -379,8 +424,10 @@ function attachAspectSubject(sample, useSampleStore, subjectModel,
 
 module.exports = {
   getNamespaceString,
```
Comment (iamigo) [inline realtime/utils.js]: getPerspectiveNamespaceString
getBotNamespaceString
initializePerspectiveNamespace
initializeBotNamespace

File: tests/realtime/realtimeUtils.js [hunk 12]
```diff
@@ -141,6 +162,12 @@ describe('tests/realtime/realtimeUtils.js, realtime utils Tests >', () => {
         .to.equal('/NA.US&EXCLUDE&EXCLUDE&EXCLUDE&EXCLUDE');
       });
 
+      it('for roomTest', () => {
+        const nspString = realtimeUtils.getNamespaceStringBots(roomTest);
+        expect(nspString)
+        .to.equal('/&' + roomTest.name);
```
Comment (iamigo) [inline tests/realtime/realtimeUtils.js]: no reason to break this onto two lines

File: tests/realtime/realtimeUtils.js [hunk 13]
```diff
@@ -132,6 +141,18 @@ describe('tests/realtime/realtimeUtils.js, realtime utils Tests >', () => {
         expect(realtimeUtils.shouldIEmitThisObj(nspString,
           looksLikeSampleObjNAUS)).to.equal(true);
       });
+
+      it('for should return true for roomTest', () => {
+        const nspString = realtimeUtils.getNamespaceStringBots(roomTest);
+        expect(realtimeUtils.shouldIEmitThisObj(nspString,
```
Comment (iamigo) [inline tests/realtime/realtimeUtils.js]: `expect(realtimeUtils.shouldIEmitThisObj(nspString, roomTest))` should all stay on one line together
`.to.equal(...)` should be on next line

File: db/helpers/common.js (hunk 6 above)
Comment (BrandonMurnane) [reply db/helpers/common.js]: i was just copying the same format that you were using  in https://github.com/salesforce/refocus/blob/master/realtime/utils.js#L37
for subjects and samples

File: realtime/utils.js (hunk 10 above)
Comment (BrandonMurnane) [reply realtime/utils.js]: bot namespace only has 2 components

**Comment by BrandonMurnane** [comment]
@iamigo is there a way to check jscs locally ?

File: realtime/utils.js (hunk 10 above)
Comment (iamigo) [reply realtime/utils.js]: This is contained within a check for subject absolutePath -- does your use case even get inside of this "if"?

File: realtime/utils.js (hunk 9 above)
Comment (BrandonMurnane) [reply realtime/utils.js]: same as above for the function name

Comment (iamigo) [reply realtime/utils.js]: > same as above for the function name

regardless, please update

**Comment by iamigo** [comment]
build failure @BrandonMurnane

File: realtime/utils.js [hunk 14]
```diff
@@ -188,49 +188,50 @@ function applyFilter(filterString, objValues) {
     /*
      * if any of the values in the objValueArr is found in the filterValueSet
      * return false
-     */
+    */
```
Comment (iamigo) [inline realtime/utils.js]: needs one more space before the asterisk -- the asterisks should line up

File: realtime/utils.js [hunk 15]
```diff
@@ -188,49 +188,50 @@ function applyFilter(filterString, objValues) {
...
   if (nspComponents.length < 2) {
     return true;
   }
 
-    /*
-     * if this is a subject object, just apply the subjcTagFilter and return
-     * the results
-     */
+  /*
+   * if this is a subject object, just apply the subjcTagFilter and return
```
Comment (iamigo) [inline realtime/utils.js]: "subjcTagFilter"?

File: realtime/utils.js [hunk 16]
```diff
@@ -175,13 +188,53 @@ function applyFilter(filterString, objValues) {
...
+   * the results
+   */
+  if (isThisSubject(obj)) {
+    return applyFilter(subjectTagFilter, obj.tags);
+  }
+
+  // apply all the filters and return the result
+  return applyFilter(aspectFilter, obj.aspect.name) &&
+      applyFilter(subjectTagFilter, obj.subject.tags) &&
```
Comment (iamigo) [inline realtime/utils.js]: too many spaces on these lines

File: realtime/utils.js [hunk 17]
```diff
@@ -211,7 +211,7 @@ function perspectiveEmit(nspComponents, obj) {
   }
 
   /*
-   * if this is a subject object, just apply the subjcTagFilter and return
+   * if this is a subject object, just apply the subjecTagFilter and return
```
Comment (iamigo) [inline realtime/utils.js]: still missing a letter here, obviously not blocking, still waiting for @shriramshankar review

Comment (BrandonMurnane) [reply realtime/utils.js]: lol sorry brain fart

File: api/v1/swagger.yaml [hunk 18]
```diff
@@ -9596,7 +9596,7 @@ paths:
           name: key
           in: path
           description: >-
-            The id the event to retrieve
+            The id the event to retrieve
```
Comment (shriramshankar) [inline api/v1/swagger.yaml]: The id of the event to retrieve?

File: db/helpers/common.js [hunk 19]
```diff
@@ -13,9 +13,12 @@
  */
 'use strict'; // eslint-disable-line strict
 
-const pub = require('../../cache/redisCache').client.pub;
```
Comment (shriramshankar) [inline db/helpers/common.js]: Actually, we want to remove all the realtime stuff from this file and only want to use what ever is under the realtime folder for pub/sub.

File: db/helpers/common.js [hunk 20]
```diff
@@ -188,7 +201,11 @@ function publishChange(inst, event, changedKeys, ignoreAttributes) {
   }
 
   if (obj[event]) {
-    pub.publish(channelName, JSON.stringify(obj));
+    if (isRoom(inst)) {
+      pubBot.publish(botChannelName, JSON.stringify(obj));
+    } else {
+      pubPerspective.publish(perspectiveChannelName, JSON.stringify(obj));
+    }
   }
```
Comment (shriramshankar) [inline db/helpers/common.js]: You probably want to make these changes to the "publishObject" method in /realTime/redisPublisher.js.

File: db/model/room.js [hunk 21]
```diff
@@ -75,8 +80,28 @@ module.exports = function room(seq, dataTypes) {
...
+          const ignoreAttributes = ['isDeleted'];
+          return common.publishChange(instance, roomEventNames.add, changedKeys,
+              ignoreAttributes);
         });
       },
+
+      afterUpdate(instance /* , opts */) {
+        if (instance.changed('settings')) {
+          if (instance.active) {
+            common.publishChange(instance, roomEventNames.upd);
```
Comment (shriramshankar) [inline db/model/room.js]: Once you have made the changes to the "publishObject" method in /realTime/redisPublisher.js, call it here.

File: realtime/setupSocketIO.js [hunk 22]
```diff
@@ -66,8 +67,18 @@ function setupNamespace(io) {
...
+        .then((rooms) => {
+          if (rooms) {
+            rooms.forEach((r) => rtUtils.initializeBotNamespace(r, io));
+            resolve(io);
+          } else {
+            const err = new ResourceNotFoundError();
+            err.resourceType = 'Room';
+            throw err;
+          }
+        });
```
Comment (shriramshankar) [inline realtime/setupSocketIO.js]: This way, if the perpsective.findAll throws an error the room namespaces will not be initialized. If this is not the intention, you might want to decouple the perspective and room initialization.

File: realtime/utils.js [hunk 23]
```diff
@@ -182,6 +195,46 @@ function applyFilter(filterString, objValues) {
   return true;
 }
 
+function perspectiveEmit(nspComponents, obj) {
```
Comment (shriramshankar) [inline realtime/utils.js]: The function description is missing.

File: realtime/utils.js [hunk 24]
```diff
@@ -182,6 +195,46 @@ function applyFilter(filterString, objValues) {
...
+  }
+
+  // apply all the filters and return the result
+  return applyFilter(aspectFilter, obj.aspect.name) &&
+    applyFilter(subjectTagFilter, obj.subject.tags) &&
+    applyFilter(aspectTagFilter, obj.aspect.tags) &&
+    applyFilter(statusFilter, obj.status);
+}
+
+function botEmit(nspComponents, obj) {
```
Comment (shriramshankar) [inline realtime/utils.js]: The function description is missing.

**Comment by shriramshankar** [comment]
Wondering if some changes also need to be done in the view layer too?

File: api/v1/swagger.yaml (hunk 18 above)
Comment (BrandonMurnane) [reply api/v1/swagger.yaml]: im not sure why its registering this as a change

File: db/model/room.js [hunk 25]
```diff
@@ -14,9 +14,14 @@
  */
 
 const constants = require('../constants');
-const u = require('../helpers/roomTypeUtils');
+const common = require('../../realtime/redisPublisher');
```
Comment (shriramshankar) [inline db/model/room.js]: can we name it "realTime" or something more relevant?

File: db/helpers/common.js [hunk 26]
```diff
@@ -137,6 +140,16 @@ function createDBLog(inst, eventType, changedKeys, ignoreAttributes) {
...
 /**
+ * A function to see if an instance is an instance of a room
+ * Checks the name from the model
+ * @param  {Object}  obj - An object instance
+ * @returns {Boolean} - returns true if the name singular is room
+ */
+function isRoom(inst) {
+  return inst.$modelOptions.name.singular === 'Room';
+}
+
```
Comment (shriramshankar) [inline db/helpers/common.js]: Can you also remove the changes related to rooms in this file?

File: realtime/setupSocketIO.js [hunk 27]
```diff
@@ -68,23 +68,25 @@ function setupNamespace(io) {
...
+        if (rooms) {
+          rooms.forEach((r) => rtUtils.initializeBotNamespace(r, io));
+          resolve(io);
+        } else {
+          const err = new ResourceNotFoundError();
+          err.resourceType = 'Room';
+          throw err;
+        }
+      });
+    })
```
Comment (shriramshankar) [inline realtime/setupSocketIO.js]: probably flatten this promise?
`.then(() => room.findAll())
.then(() => { if(rooms) ......})`

Comment (shriramshankar) [reply realtime/setupSocketIO.js]: Can you remove the else blocks too, 83 to 87 and 71 to 75. I think we will never hit these lines.

**Comment by BrandonMurnane** [comment]
@shriramshankar implemented changes and fixed tests. anything else you need ?

**Comment by BrandonMurnane** [comment]
This pr requires 2 environment variable changes. Current pubsub needs to be changed to REDIS_PUBSUB_PERSPECTIVES. Also a new redis instance and new variable needs to be added that has that address REDIS_PUBSUB_BOTS. @iamigo i know refocus has a release manager is there a specific way we should be communicating this information.

**Comment by iamigo** [comment]
Both of these can be set up in advance in Heroku--I'll do that now. After the release is finished and we know we're not rolling back or anything, we just need to clean up after ourselves and delete the unused env var from the Heroku config settings page.

**Comment by shriramshankar** [comment]
@BrandonMurnane  had a question for you in one of my previous comments 
"Wondering if some changes also need to be done in the view layer too?"

**Comment by BrandonMurnane** [comment]
@shriramshankar sorry that i missed that. Do tests not pick this up ? im not sure how the view layer is working if u have any suggestions how to check if it needs changes would be helpful.

**Comment by shriramshankar** [comment]
@BrandonMurnane Do the real-time events work locally? For the real-time events to reach the client(browser), you will have to force the client to listen on a namespace. Check out "setupSocketIOClient" in view/perspective/app.js.

**Comment by BrandonMurnane** [comment]
@shriramshankar in the setupsocketclient in that file u call getNamespaceString from the utils file but it doesnt exist there.

**Comment by shriramshankar** [comment]
@BrandonMurnane it is in "view/utils.js"
